packages/react-instantsearch-core/src/widgets/__tests__/DynamicWidgets.test.tsx

```diff
@@ -0,0 +1,231 @@
+import { render } from '@testing-library/react';
+import React from 'react';
+import { connectHierarchicalMenu, connectRefinementList } from '../..';
+import DynamicWidgets from '../DynamicWidgets';
+import InstantSearch from '../InstantSearch';
+
+const EMPTY_RESPONSE = {
+  results: [
+    {
+      hits: [],
+      nbHits: 0,
+      page: 0,
+      nbPages: 0,
+      hitsPerPage: 20,
+      exhaustiveNbHits: true,
+      query: '',
+      queryAfterRemoval: '',
+      params:
+        'highlightPreTag=%3Cais-highlight-0000000000%3E&highlightPostTag=%3C%2Fais-highlight-0000000000%3E&query=&facets=%5B%5D&tagFilters=',
+      index: 'instant_search',
+      processingTimeMS: 2,
+    },
+  ],
+};
+
+const createSearchClient = () => ({
+  search: jest.fn(() => Promise.resolve(EMPTY_RESPONSE)),
+  searchForFacetValues: jest.fn(() => Promise.resolve({})),
+});
+
+const RefinementList = connectRefinementList(
+  ({ attribute }) => `RefinementList(${attribute})`
+);
+
+const HierarchicalMenu = connectHierarchicalMenu(
+  ({ attributes }) => `HierarchicalMenu(${attributes.join(',')})`
+);
+
+describe('DynamicWidgets', () => {
```

Comment:
Perhaps add a test with deeply nested components so that we can align and track that behavior?